cpp/src/decisiontree/batched-levelalgo/kernels.cuh

```diff
@@ -264,13 +264,77 @@ __device__ OutT* alignPointer(InT input) {
     raft::alignTo(reinterpret_cast<size_t>(input), sizeof(OutT)));
 }
 
+// 32-bit FNV1a hash
+// Reference: http://www.isthe.com/chongo/tech/comp/fnv/index.html
+const uint32_t fnv1a32_prime = uint32_t(16777619);
+const uint32_t fnv1a32_basis = uint32_t(2166136261);
+
+DI uint32_t fnv1a32(uint32_t hash, uint32_t txt) {
+  hash ^= (txt >> 0) & 0xFF;
+  hash *= fnv1a32_prime;
+  hash ^= (txt >> 8) & 0xFF;
+  hash *= fnv1a32_prime;
+  hash ^= (txt >> 16) & 0xFF;
+  hash *= fnv1a32_prime;
+  hash ^= (txt >> 24) & 0xFF;
+  hash *= fnv1a32_prime;
+  return hash;
+}
+
+/**
+ * @brief For a given values of (treeid, nodeid, seed), this function generates
+ *        a unique permutation of [0, N - 1] values and returns 'k'th entry in
+ *        from the permutation.
+ * @return The 'k'th value from the permutation
+ * @note This function does not allocated any temporary buffer, all the
+ *       necessary values are recomputed.
+ */
+template <typename IdxT>
+DI IdxT select(IdxT k, IdxT treeid, uint32_t nodeid, uint64_t seed, IdxT N) {
+  __shared__ int blksum;
+  uint32_t pivote_hash;
```

Comment:
(minor/picky): should this be `pivot_hash`?

Reply:
Good catch, I'll correct the spelling.

Reply:
Done.